source/common/config/watch_map.cc

```diff
@@ -93,10 +93,36 @@ void WatchMap::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>
   }
 }
 
+// For responses to on-demand requests, replace the original watch for an alias
+// with one for the resource's name
+void WatchMap::convertAliasWatchToNameWatch(const envoy::api::v2::Resource& resource) {
+  absl::flat_hash_set<Watch*> watches_to_update;
+  for (const auto& alias : resource.aliases()) {
+    const auto interested_watches = watch_interest_.find(alias);
+    if (interested_watches != watch_interest_.end()) {
+      for (const auto& interested_watch : interested_watches->second) {
+        watches_to_update.insert(interested_watch);
+      }
+      break;
+    }
+  }
+  for (const auto& watch : watches_to_update) {
+    updateWatchInterest(watch, {resource.name()});
```

Comment:
I'm worried about your use of `updateWatchInterest` here. Is it the case that it will only ever return {empty, empty}? If not, you'll need to get those changes in subscription interest communicated out to GrpcMux, I think, or else some part of the xDS process might just start ignoring these freshly resolved resource names.

Reply:
Ugh, yes, you are right. Changed to update subscription too:
https://github.com/envoyproxy/envoy/pull/8617/files#diff-050ecb76fd3d458fd208a27e7c1a4e23R66. Please note I relaxed access to NewGrpcMuxImpl::subscriptions_ -- needed that for tests.